Sources/NIO/EventLoopFuture.swift

```diff
@@ -957,6 +957,43 @@ extension EventLoopFuture {
     }
 }
 
+extension EventLoopFuture {
+    /// Returns a new `EventLoopFuture` that succeeds when all of the provided `EventLoopFuture`s complete.
+    /// The new `EventLoopFuture` will contain an array of results, maintaining ordering for each of the `EventLoopFuture`s.
+    ///
+    /// The returned `EventLoopFuture` always succeeds, regardless of any failures from the waiting futures.
+    ///
+    /// If it is desired to flatten them into a single `EventLoopFuture` that fails on the first `EventLoopFuture` failure,
+    /// use one of the `reduce` methods instead.
+    /// - Parameter futures: An array of homogenous `EventLoopFuture`s to gather results from.
+    /// - Returns: A new `EventLoopFuture` with all the results of the provided futures.
+    public static func whenAllComplete<InputValue>(_ futures: [EventLoopFuture<InputValue>],
+                                                   eventLoop: EventLoop) -> EventLoopFuture<[Result<InputValue, Error>]> {
+        return eventLoop.makeSucceededFuture(result: [])
+            .flatMap { _ in
+                let promise = eventLoop.makePromise(of: [Result<InputValue, Error>].self)
+
+                var remainingCount = futures.count
+                var results: [Result<InputValue, Error>?] = .init(repeating: nil, count: futures.count)
```

Comment:
Oh, yes, I meant to come back to this.

The `compactMap { $0 }` you need to perform at the bottom here is a bit awkward: it forces a new array allocation, and we happen to know analytically that there should never be nils in this array. So I think I'd propose rewriting this to remove the optionality, and instead initialising the array with a bunch of `Result`s in the `Error` state with an appropriate `Error`. That way the optionals are never needed.

Reply:
Is there an existing data structure for EventLoopFuture errors, or should I create a new one?